server/stream.go

```diff
@@ -79,11 +81,27 @@ type Stream struct {
 	consumers map[string]*Consumer
 	config    StreamConfig
 	created   time.Time
+	ddmap     map[string]*ddentry
+	ddarr     []*ddentry
+	ddindex   int
+	ddtmr     *time.Timer
 }
 
+// JSPubId is used for identifying published messages and performing de-duplication.
+const JSPubId = "Msg-Id"
```

Comment:
I don't think we make it a config item. And I went back and forth on it. I did not like `Js-Msg-Id` but we could go with that.

Reply:
ok, but we need to accept it case insensitive as per http spec

Reply:
Headers have a canonical name that is the way it is above. Uppercase for each word, words can be separated by `-`

Reply:
Yes, so for the storage in the map, we need to decide on converting to lower or upper prior that.
By the way, this is just stored in memory as of now or did I miss it? Meaning would that survive a server restart?

Reply:
Good point, it is saved but not in that form per se, would need to re-populate. Will make that test later this am and fix. Good catch.

Reply:
About case sensitive: our own ADR (https://github.com/nats-io/nats-server/blob/master/doc/adr/0004-nats-headers.md) says:
```
Each header field consists of a case-insensitive field name followed by a colon (:), optional leading whitespace, the field value, and optional trailing whitespace.
```

Reply:
Or: https://tools.ietf.org/html/rfc7230#section-3.2
```
Each header field consists of a case-insensitive field name followed
   by a colon (":"), optional leading whitespace, the field value, and
   optional trailing whitespace.
```

Reply:
https://golang.org/pkg/net/textproto/#CanonicalMIMEHeaderKey

Reply:
Those headers will come from clients and are not necessarily Go. If you think they must be set like that, then the ADR should be updated...

Reply:
Yes they should take that form, do not want to do multiple searches, already had to re-architect because generic hoisting to http style headers was too slow.

Reply:
I think its important that if we're going to tell a http header story then its a minimum wtf/minute story, so should be case insensitive.

Reply:
Agreed. Some client libraries may use some http libraries that will adhere to the http spec.

Reply:
All client libs that do HTTP do the canonical form and that is what we will search for an require in the server. Too expensive and not worth it to lowercase copies of headers. Purposely trying to avoid additional allocations and copies etc.